api/institutions/authentication.py

```diff
@@ -0,0 +1,68 @@
+import json
+
+import jwe
+import jwt
+
+from rest_framework.exceptions import AuthenticationFailed
+from rest_framework.authentication import BaseAuthentication
+
+from api.base import settings
+from website.models import Institution
+from framework.auth import get_or_create_user
+
+
+class InstitutionAuthentication(BaseAuthentication):
+    media_type = 'text/plain'
+
+    def authenticate(self, request):
+        try:
+            payload = jwt.decode(
+                jwe.decrypt(request.body, settings.JWE_SECRET),
+                settings.JWT_SECRET,
+                options={'verify_exp': False},
+                algorithm='HS256'
+            )
+        except (jwt.InvalidTokenError, KeyError, ValueError, TypeError):
+            raise AuthenticationFailed
+
+        # The JWT `data` payload is expected in the following structure.
+        #
+        # {"provider": {
+        #     "idp": "https://login.circle.edu/idp/shibboleth",
+        #     "id": "CIR",
+        #     "user": {
+        #         "middleNames": "",
+        #         "familyName": "",
+        #         "givenName": "",
+        #         "fullname": "Circle User",
+        #         "suffix": "",
+        #         "username": "user@circle.edu"
+        #     }
+        # }}
+        data = json.loads(payload['data'])
+        provider = data['provider']
+
+        institution = Institution.load(provider['id'])
+        if not institution:
+            raise AuthenticationFailed('Invalid institution id specified "{}"'.format(provider['id']))
+
+        username = provider['user']['username']
+        fullname = provider['user']['fullname']
+
+        user, created = get_or_create_user(fullname, username)
```

Comment:
You might need to lowercase `username` here; it doesn't look like `get_or_create_user` will lowercase it, so it might be possible to create duplicate users with differently-cased usernames.